src/query-builder/json-path-builder.ts

```diff
@@ -0,0 +1,276 @@
+   * )
+   * ```
+   *
+   * The generated SQL (PostgreSQL):
+   *
+   * ```sql
+   * select "experience"->0->'role' as "first_role" from "person"
+   * ```
+   *
+   * You can use `'last'` to access the last element of the array in MySQL:
+   *
+   * ```ts
+   * db.selectFrom('person').select(eb =>
+   *   eb.ref('nicknames', '->$').at('last').as('last_nickname')
+   * )
+   * ```
+   *
+   * The generated SQL (MySQL):
+   *
+   * ```sql
+   * select `nicknames`->'$[last]' as `last_nickname` from `person`
+   * ```
+   *
+   * Or `'#-1'` in SQLite:
+   *
+   * ```ts
+   * db.selectFrom('person').select(eb =>
+   *   eb.ref('nicknames', '->>$').at('#-1').as('last_nickname')
+   * )
+   * ```
+   *
+   * The generated SQL (SQLite):
+   *
+   * ```sql
+   * select "nicknames"->>'$[#-1]' as `last_nickname` from `person`
+   * ```
+   */
+  at<
+    I extends any[] extends O ? number | 'last' | `#-${number}` : never,
+    O2 = null | NonNullable<NonNullable<O>[keyof NonNullable<O> & number]>
+  >(
+    index: `${I}` extends `${any}.${any}` | `#--${any}` ? never : I
+  ): TraversedJSONPathBuilder<S, O2> {
+    return this.#createBuilderWithPathLeg('ArrayLocation', index)
+  }
+
+  /**
+   * Access a property of a JSON object.
+   *
+   * If a field is optional, the resulting type will be nullable.
+   *
+   * See also {@link at} to access elements of JSON arrays.
+   *
+   * ### Examples
+   *
+   * ```ts
+   * db.selectFrom('person').select(eb =>
+   *   eb.ref('address', '->').key('city').as('city')
+   * )
+   * ```
+   *
+   * The generated SQL (PostgreSQL):
+   *
+   * ```sql
+   * select "address"->'city' as "city" from "person"
+   * ```
+   *
+   * Going deeper:
+   *
+   * ```ts
+   * db.selectFrom('person').select(eb =>
+   *   eb.ref('profile', '->$').key('website').key('url').as('website_url')
+   * )
+   * ```
+   *
+   * The generated SQL (MySQL):
+   *
+   * ```sql
+   * select `profile`->'$.website.url' as `website_url` from `person`
+   * ```
+   *
+   * Combined with {@link at}:
+   *
+   * ```ts
+   * db.selectFrom('person').select(eb =>
+   *   eb.ref('profile', '->').key('addresses').at(0).key('city').as('city')
+   * )
+   * ```
+   *
+   * The generated SQL (PostgreSQL):
+   *
+   * ```sql
+   * select "profile"->'addresses'->0->'city' as "city" from "person"
+   * ```
+   */
+  key<
+    K extends any[] extends O
+      ? never
+      : O extends object
+      ? keyof NonNullable<O> & string
+      : never,
+    O2 = undefined extends O
+      ? null | NonNullable<NonNullable<O>[K]>
+      : null extends O
+      ? null | NonNullable<NonNullable<O>[K]>
+      : NonNullable<O>[K]
+  >(key: K): TraversedJSONPathBuilder<S, O2> {
+    return this.#createBuilderWithPathLeg('Member', key)
+  }
+
+  #createBuilderWithPathLeg(
+    legType: JSONPathLegType,
+    value: string | number
+  ): TraversedJSONPathBuilder<any, any> {
+    const legNode = JSONPathLegNode.create(
+      legType,
+      RawNode.createWithSql(String(value))
```

Comment:
I think this should be a `ValueNode.createImmediate` instead. That way it's somewhat transformable. Or am I missing something here?

Reply:
I thought so too, but `ValueNode.createImmediate` wraps strings with `'`s and JSON path should be `->'$.key[last]'`.
Maybe we should create a `JSONPathLegValueNode`?